### ISSUE TITLE
Add VISIBLE and REQUIRED support for Related Form Control feature

### ISSUE THREAD

**Issue by anteburazer** [udos86/ng-dynamic-forms#253]
Hi Udo

Very nice work with this library.

Is it possible to extend **Related Form Control** (https://github.com/udos86/ng2-dynamic-forms#related-form-controls) with **VISIBLE** and **REQUIRED** actions? For now there's only **DISABLED** action, as I can see.

 **VISIBLE** action shows/hides an element if some other element is set to particular value.
 **REQUIRED** action makes the element required if some other element is set to particular value.

Thanks

**Comment by udos86** [comment]
@anteburazer Thanks for the kind words!

Your'e absolutely right. At the moment you can only relate form controls concerning activation state (DISABLED / ENABLED). This is the part of the library I'm not really satisfied with. I'd like to refactor and improve that feature but I'm lacking a bit of time right now. However, I'll try my best to add support for this in the near future.

**Comment by minuz** [comment]
Yes, that is a very desirable feature! I was about to suggest this as well.

**Comment by udos86** [comment]
@minuz I'm still considering a total refactoring of the related form controls feature. I'll check how much effort it would cost to implement the requested properties However at first I'd like to pave the way for all revelant form controls (FileUpload, Datepicker, etc.).

**Comment by friendsofdevs** [comment]
Extremely helpful feature! Now I'm using complex forms (with 3-4 of controls) and it will be better with VISIBLE property. My case is:  formgroup control it's a select box with values: Exact, Less than, Great than, In between. So if someone chooses 'In between' form shows two fields for range, otherwise only one.

**Comment by minuz** [comment]
@udos86, sounds good! 

The refactoring seems almost inevitable. I've sort of been needing to have different validations based on the relations as well. With the current structure, I'm not sure if that's possible. 

Let's see when I get some spare time to help you out. This library will be extremely useful with all this functionality so I would gladly help you.

**Comment by minuz** [comment]
Just wanted to let people know that for now, if you really need to hide controls, you can sort of hack your way by using change event as follows: 

```HTML
<dynamic-form-bootstrap-control *ngFor="let controlModel of formModel" 
         [controlGroup]="form" 
         [model]="controlModel" 
         id="container-{{controlModel.id}}" 
         (change)="onValueChange($event)">
</dynamic-form-bootstrap-control>
```

Then you will have the following available:

```javascript
onValueChange(event: DynamicFormControlEvent) {
    if (event.model.type === 'SELECT') {
        //Hide/Show your control using jQuery
        switch (event.control.value) {
            case 'Option1':
                $('#inputContainerId').show();
                break;
            default:
                $('#inputContainerId').hide();
                $('#inputId').val('');
                break;
        }
    }
}
```

**Comment by nomi-ramzan** [comment]
Do anybody have any idea about default hide on initialization ? then based on value change we can show related section But i am trying to implement it on `DynamicFormGroupModel`
Thank you

**Comment by minuz** [comment]
@nomi-ramzan, if you read this thread, you see that currently there's no default implementation that will allow you to do this out-of-the-box. My previous comment shows how to hack it around using jQuery as shortcut. However, it could also be achieved by using the angular ```elementRef```. The key element is to use the built-in ```change``` event to handle whilst rendering the DOM.

**Comment by nomi-ramzan** [comment]
@minuz But your solution is kind of hard-coded and hard to implement for complex forms. Because in this way `$('#inputContainerId')` many keys i have to manage. But I will read about `elementRef`. Thank you for your solution.

**Comment by Palad1NPT** [comment]
#335 

Added a hack to be able to hide and show controls using relation.
Limitations in the same post.

**Comment by matteonovelli** [comment]
Maybe the easiest way is to bind hidden attribute with controlModel's disabled property 


``` html
<dynamic-form-bootstrap-control [group]="formGroup"
                                    [model]="controlModel"
                                    #componentRef
                                    [hidden]="controlModel.disabled">
 </dynamic-form-bootstrap-control>
```

**Comment by nomi-ramzan** [comment]
Hi @matteonovelli I just tried your solution but sorry its not working for me. Can you please provide more details or any working example will be great.

**Comment by matteonovelli** [comment]
Hi @nomi-ramzan here is a working example:

```html
<form [formGroup]="formGroup">
  <ng-container *ngFor="let controlModel of formModel; trackBy: trackByForm">
      <dynamic-form-bootstrap-control [group]="formGroup"
                                                                [model]="controlModel"
                                                                [hidden]="controlModel.disabled">
      </dynamic-form-bootstrap-control>
   </ng-container>
</form>
```

```typescript
import { Component, Input, OnChanges, SimpleChange, SimpleChanges } from '@angular/core';
import { DynamicFormControlModel, DynamicFormService } from '@ng2-dynamic-forms/core';
import { FormGroup } from '@angular/forms';

@Component({
  selector: 'of-dynamic-form',
  templateUrl: './dynamic-form.component.html'
})
export class DynamicFormComponent implements OnChanges {
  @Input() formModel: DynamicFormControlModel[] = [];
  formGroup: FormGroup;

  constructor(private formService: DynamicFormService) { }

  ngOnChanges(changes: SimpleChanges) {
    const formModel = changes['formModel'] as SimpleChange;
    if (formModel && formModel.currentValue !== undefined) {
      this.formGroup = this.formService.createFormGroup(this.formModel);
    }
  }

  trackByForm(index: number, controlModel: DynamicFormControlModel) {
    return controlModel.id;
  }
}
```

The "magic" happens because this binding:  [hidden]="controlModel.disabled". Also, as can you see in the typescript, I recreate the form every time the formModel changes

**Comment by bdparrish** [comment]
I would add to the feature  enhancement with a `RELOAD` option.  I would also be willing to put in cycles if you can point me in the right direction to add these options before the refactor is done.

**Comment by joanllenas** [comment]
Hi @udos86 awesome work with this library 💯
Is there any reason why you are not accepting this [PR](https://github.com/udos86/ng-dynamic-forms/pull/547), which partially solves this issue?
The show/hide workaround route could be an option for us but I wanted to ask first.

**Comment by udos86** [comment]
@joanllenas Thank you!

> Is there any reason why you are not accepting this PR, which partially solves this issue?

Yes. There's a major refactoring of the related form controls feature on my roadmap.

I need to have another look but the provided PR is moving in the wrong direction of what I have already been thinking of.

However some other stuff needs to be done first before I'll go on with this.

**Comment by z0m1x** [comment]
Thank you for your work. I also think that this is a very desirable feature. I thought of the html markup 
`<ng-template>`
And if this is added maybe only the controls will be hidden, what if one are using ng templates with insert and remove buttons, those would be great to make hidden also from the module(.ts) markup. 

This kind of leads me to my next point. Now its seems that i need to use ngtemplates in my form to handle for example insert/remove, file upload. I would like to have the form complete rendered from data in a NoSQL database, but then i guess logic need to me moved to the DynamicForm model.

```
    // new DynamicInputModel({

    //     id: "lastName",
    //     label: "Last Name",
    //     placeholder: "Last Name",
    //     validators: {
    //         required: null
    //     },
    //     errorMessages: {
    //         required: "{{ label }} is required"
    //     },
           Click: [ {"action": "remove", "name": "removeItem", "params": ""}, {"action": "insert", "name": "removeItem", "params": ""}],
     // }),
```

**Comment by udos86** [comment]
Basis support for hiding form controls will be available in `5.4.1`.

Merry Christmas!

**Comment by z0m1x** [comment]
Hi, saw the new commits in dev branch and now it semse like you published 5.4.1.
However i wonder if i'm doing something wrong.
I'm trying to set previously string "DISABLED" to "HIDDEN" or "VISIBLE".
It looks like my **NG Dynamic form control model** .ts have not been updated after npm update

My project package.json
```
    "@ng-dynamic-forms/core": "^5.4.1",
    "@ng-dynamic-forms/ui-primeng": "^5.4.1",
    "aws-sdk": "^2.171.0",
```

**NG Dynamic form control model** (from your commit to dev branch)
```
   dynamic-form-control-model.d.ts
    _disabled: boolean;
    disabledUpdates: Subject<boolean>;
    errorMessages: DynamicValidatorsConfig | null;
```

```
+        this.disabledUpdates.subscribe(disabled => this.disabled = disabled);
  		  
          this.hidden = typeof config.hidden === "boolean" ? config.hidden : false;		          this.hidden = typeof config.hidden === "boolean" ? config.hidden : false;
          this.hiddenUpdates = new Subject<boolean>();		          this.hiddenUpdates = new Subject<boolean>();
 -        this.hiddenUpdates.subscribe((value: boolean) => this.hidden = value);		 +        this.hiddenUpdates.subscribe(hidden => this.hidden = hidden);
```

This changes above does not exist in my class after npm update. Do you or someone know why?

Happy hollidays! (or Marry Christmas :P )

**Comment by z0m1x** [comment]
@udos86 Am i understanding this correct? This issue is regarding hidden field (A) and not as an action (B)?

**Comment by udos86** [comment]
@z0m1x Sorry for the late reply!

> Am i understanding this correct? This issue is regarding hidden field (A) and not as an action (B)?

Yes, currently there's no action to for hiding a form control. 

But at least you're now able to set the hidden state via the corresponding property on the model.

To be continued... :-)

**Comment by z0m1x** [comment]
@udos86 

Okey, so as i see it, this would be great to have. Also to consider that when used, the label also gets hidden (or add an 'add to parent class' when condition is met like my last example).

This might be a bit off topic but, in "ui-primeng.es.js" there is a template markup assigned to decorators property for DynamicPrimeNGFormControlComponent.

Where: ngSwitchCase=10 is regarding input text fields. However when using InputType="file" this is used also. I did a fast workaround with 
`<div *ngIf=\"model.inputType == 'file';then afileinput else aninputtext\">here is ignored</div>...`

then i have the two different markup for input text and input file.
Because this is primeng i'm using their file template:
`<ng-template #afileinput><ng-container *ngSwitchCase=\"10\"><p-fileUpload"...`

Now i know there is a 'DynamicFileUploadModel' component.

but it seems like there is no corresponding template for this, no "ngSwitchCase=XX". Am i missing something? (you push for using input text with InputType="file" on the start git repo page)

I think it would be better to add this, and use 'DynamicFileUploadModel' for file uploads/selects and when using primeng, have like *ngSwitchCase=xx then 
`<p-fileUpload"...`

This might even be more OT:
I would like to (for now) extend and change decorators property for 'DynamicPrimeNGFormControlComponent' that is defined in "ui-primeng.es.js". How could i do that? I'm thinking on initialize of the app change the decorators and thereby the template.

Br :)

**Comment by boyski33** [comment]
```
  new DynamicInputModel({
    id: 'someId',
    label: 'Input Field 1',
    relation: [
      {
        action: 'HIDDEN',
        when: [
          {id: 'someOtherId', value: 'value123'}
        ]
      }
    ]
  }),

```

If I'm understanding correctly, this is supposed to be the end result, right: the code above should hide `someId` if the value of `someOtherId` is `value123`?

**Comment by udos86** [comment]
will finally be available in `8.0.0`

**Comment by anuj-pandit** [comment]
Is there any example of this ? Was wondering if we can do something like below?

E.x. Have 3 fields. value of field 1 controls visibility of field 2 and 3 and whichever field is visible need to be mandatory. Whats the cleanest way of doing this. Also, I want to avoid re-executing the display rules.
 
this.form = this._formBuilder.group({
				field1: ['', Validators.required],
				field2:['', [Validators.required,Validator.Visible]  ],
				field3: ['', [Validators.required,Validator.Visible] ]
			}
